### PR TITLE
Add elementwise checks to kernel generator

### PR DESCRIPTION
**Pull request by t4c1**
## Summary
Implements `check_cl` - a way to do checks on opencl arguments.

The interface is a bit weird:
```
check_cl("function", "m1", m1_cl, "positive") = m1_cl > 0;
```
However, it allows us to perform checks on inputs in the same OpenCL kernel as computations:
```
results(m2_cl, check_cl("test", "m1", m1_cl, "positive"))
               = expressions(m1_cl * 2 + 1, m1_cl > 0);
```
This is important for performance - following the whole kernel generator logic. It is also very general, so we do not need a new function for every different check.

I also modified `value_type_t` metaprogram to also work on kernel generator expressions.

## Tests
check_cl is tested.

## Side Effects
Kernels that contain checks are not allowed to run asynchronously. Check needs to wait for the kernel to finish so it can check the result and throw if necessary - throwing can not happen asynchronously.

## Release notes
Added support for elementwise checks to kernel generator

## Checklist

- [x] Math issue #1342 

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/opencl/kernel_generator/check_cl.hpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+namespace stan {
+namespace math {
+
+/** \addtogroup opencl_kernel_generator
+ *  @{
+ */
+
+/**
+ * Represents a check in kernel generator expressions. When a check is assigned
+ * an expression it will throw if any value in the result of the expression is
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_generator/check_cl.hpp:20]: *to an expression

File: stan/math/opencl/kernel_generator/check_cl.hpp [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  T arg_;
+  const char* function_;
+  const char* err_variable_;
+  const char* must_be_;
+
+ public:
+  /**
+   * Constructor.
+   * @param function function name (for error messages)
+   * @param err_variabhle variable name (for error messages)
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_generator/check_cl.hpp:46]: err_variable* 

also in the signature below

File: stan/math/opencl/kernel_generator/check_cl.hpp (hunk 1 above)
Comment (t4c1) [reply stan/math/opencl/kernel_generator/check_cl.hpp:20]: I think this is correct as it is.

Comment (rok-cesnovar) [reply stan/math/opencl/kernel_generator/check_cl.hpp:20]: Yes, I am a fool. Its check = expression.

**Pull request merged**